library/std/src/collections/hash/set.rs

```diff
@@ -1802,6 +1850,414 @@ where
+    ///
+    /// # Examples
+    ///
+    /// ```
+    ///  use std::collections::hash_set::{Entry, HashSet};
+    ///  use std::rc::Rc;
+    ///
+    ///  let mut set: HashSet<Rc<String>> = HashSet::new();
+    ///  let key_one = Rc::new("Stringthing".to_string());
+    ///  let key_two = Rc::new("Stringthing".to_string());
+    ///
+    ///  set.insert(key_one.clone());
+    ///  assert!(Rc::strong_count(&key_one) == 2 && Rc::strong_count(&key_two) == 1);
+    ///
+    ///  match set.entry(key_two.clone()) {
+    ///      Entry::Occupied(entry) => {
+    ///          let old_key: Rc<String> = entry.replace();
+    ///          assert!(Rc::ptr_eq(&key_one, &old_key));
+    ///      }
+    ///      Entry::Vacant(_) => panic!(),
+    ///  }
+    ///
+    ///  assert!(Rc::strong_count(&key_one) == 1 && Rc::strong_count(&key_two) == 2);
+    ///  assert!(set.contains(&"Stringthing".to_owned()));
+    /// ```
+    #[inline]
+    #[unstable(feature = "hash_set_entry", issue = "60896")]
+    pub fn replace(self) -> T {
+        self.base.replace()
+    }
+}
+
+impl<'a, T, S> VacantEntry<'a, T, S> {
+    /// Gets a reference to the value that would be used when inserting
+    /// through the `VacantEntry`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashSet;
+    ///
+    /// let mut set: HashSet<&str> = HashSet::new();
+    /// assert_eq!(set.entry("poneyland").get(), &"poneyland");
+    /// ```
+    #[inline]
+    #[unstable(feature = "hash_set_entry", issue = "60896")]
+    pub fn get(&self) -> &T {
+        self.base.get()
+    }
+
+    /// Take ownership of the value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::hash_set::{Entry, HashSet};
+    ///
+    /// let mut set: HashSet<&str> = HashSet::new();
+    ///
+    /// match set.entry("poneyland") {
+    ///     Entry::Occupied(_) => panic!(),
+    ///     Entry::Vacant(v) => assert_eq!(v.into_value(), "poneyland"),
+    /// }
+    /// ```
+    #[inline]
+    #[unstable(feature = "hash_set_entry", issue = "60896")]
+    pub fn into_value(self) -> T {
+        self.base.into_value()
+    }
+
+    /// Sets the value of the entry with the VacantEntry's value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashSet;
+    /// use std::collections::hash_set::Entry;
+    ///
+    /// let mut set: HashSet<&str> = HashSet::new();
+    ///
+    /// if let Entry::Vacant(o) = set.entry("poneyland") {
+    ///     o.insert();
+    /// }
+    /// assert!(set.contains("poneyland"));
+    /// ```
+    #[inline]
+    #[unstable(feature = "hash_set_entry", issue = "60896")]
+    pub fn insert(self)
+    where
+        T: Hash,
+        S: BuildHasher,
+    {
+        self.base.insert();
+    }
+
+    #[inline]
+    fn insert_entry(self) -> OccupiedEntry<'a, T, S>
+    where
+        T: Hash,
+        S: BuildHasher,
+    {
+        todo!()
```

Comment:
@cuviper @the8472 this needs to be public in hashbrown. Is it ok if I do that for `rustc-internal-api`?

Reply:
See what @Amanieu thinks as the `hashbrown` maintainer.

Reply:
I would prefer that is be added as a public API in hashbrown too.

Reply:
What would be the difference between `insert` and `insert_entry`? It kind of seems to me like those are the same API? Please correct me if I'm wrong, but otherwise I'd propose deprecating `insert` and making `insert_entry` public in hashbrown and then just having `insert_entry -> OccupiedEntry` in the stdlib.

https://doc.rust-lang.org/std/collections/hash_map/struct.VacantEntry.html#method.insert
The hashmap needs both `insert` and `insert_entry` because one returns a reference to the value while the other to the entry. But for hashset there is no value, so we don't need insert at all.

Thoughts?

Reply:
@Amanieu thoughts on the API changes I proposed?

Reply:
I think we could just change `insert` to always return an `OccupiedEntry` and remove `insert_entry`.

Reply:
It's fine to make breaking changes to hashbrown, it just means that a major version bump is needed.

Reply:
Sounds good, done! https://github.com/rust-lang/hashbrown/pull/495

Reply:
While reviewing the entry API, I noticed a subtle issue. `insert_entry` will consume the key in `VacantEntry` to insert it into the set. This will result in it returning an `OccupiedEntry` with key set to `None`. This will then panic if you try to use `OccupiedEntry::replace` since there is no search key to replace the original key with.

Reply:
Looks like map has the same issue: https://github.com/rust-lang/hashbrown/blob/4c824c548e85d75f95e0dfcee09c62c5203f9a75/src/map.rs#L5702

Shouldn't replace do nothing in this case? We know it already has the key we just inserted so there's nothing to replace.

Reply:
Currently it will panic because `replace_key` will unwrap the inner option: https://github.com/rust-lang/hashbrown/blob/4c824c548e85d75f95e0dfcee09c62c5203f9a75/src/map.rs#L5539

I'm currently leaning towards just removing `OccupiedEntry::replace`.

Reply:
See https://github.com/rust-lang/hashbrown/pull/535.

Reply:
I have no opinion on removal, but my point was that I believe it'd be correct to just check for is_none and then short circuit/do nothing.